[CL-3184] Tenant template robustness

adessy opened the pull request:
# Changelog
- Refactoring of the tenant template system. Main modifications:
  - Templates are now self-contained and store all the uploads (images and documents) along with the serialized models.
  - Uploads are moved directly from S3 to S3 (without local download) for both the creation and application of templates. This also means that the system is tightly coupled with S3.

# TODO

- [x] ~Rework `Base` into `BaseWithTimestamps` (instead of including `Core` directly in some serializers)~
- [x] Relax `validate :admin_publication_must_exist` and `set_admin_publication` in project.rb
- [x] (optimization) skip check for `:file` presence in `event_file.rb` and `phase_file.rb` (it triggers a download).
- [x] comment and refactor the TenantSerializer
- [x] is the template-verification step still useful?
- [x] rewrite `templates:verify`
- [x] test test test test test
- [x] remove namespace where unnecessary (e.g. MultiTenancy::Templates::TenantDeserializer)

adessy committed:
[CL-3184] Add higher-level S3 utilities

adessy committed:
[CL-3184] Implement base serializer for tenant-template resources

adessy committed:
[CL-3184] Add template serializer for areas

adessy committed:
[CL-3184] Add template serializer for custom forms

adessy committed:
[CL-3184] Add template serializer for content builder layouts

adessy committed:
[CL-3184] Add template serializer for content builder layout images

adessy committed:
[CL-3184] Add template serializer for home pages

adessy committed:
[CL-3184] Add template serializer for text images

adessy committed:
[CL-3184] Add template serializer for custom field options

adessy committed:
[CL-3184] Add template serializer for topics

adessy committed:
[CL-3184] Add template serializer for project-folder images

adessy committed:
[CL-3184] Add template serializer for project-folder files

adessy committed:
[CL-3184] Add template serializer for project files

adessy committed:
[CL-3184] Add template serializer for project images

adessy committed:
[CL-3184] Add template serializer for project input topics

adessy committed:
[CL-3184] Add template serializer for phase files

adessy committed:
[CL-3184] Add template serializer for participation contexts

adessy committed:
[CL-3184] Add template serializer for email campaigns

adessy committed:
[CL-3184] Add template serializer for baskets

adessy committed:
[CL-3184] Add template serializer for events

adessy committed:
[CL-3184] Add template serializer for event files

adessy committed:
[CL-3184] Add template serializer for groups

adessy committed:
[CL-3184] Add template serializer for groups-projects

adessy committed:
[CL-3184] Add template serializer for permissions

adessy committed:
[CL-3184] Add template serializer for group permissions

adessy committed:
[CL-3184] Add template serializer for memberships

adessy committed:
[CL-3184] Add template serializer for static pages

adessy committed:
[CL-3184] Add template serializer for nav-bar items

adessy committed:
[CL-3184] Add template serializer for static-page file

adessy committed:
[CL-3184] Add template serializer for idea statuses

adessy committed:
[CL-3184] Add template serializer for baskets-ideas

adessy committed:
[CL-3184] Add template serializer for idea files

adessy committed:
[CL-3184] Add template serializer for idea images

adessy committed:
[CL-3184] Add template serializer for idea phases

adessy committed:
[CL-3184] Add template serializer for ideas-topics

adessy committed:
[CL-3184] Add template serializer for initiative statuses

adessy committed:
[CL-3184] Add template serializer for initiatives

adessy committed:
[CL-3184] Add template serializer for initiative status changes

adessy committed:
[CL-3184] Add template serializer for areas-initiatives

adessy committed:
[CL-3184] Add template serializer for initiative files

adessy committed:
[CL-3184] Add template serializer for initiative images

adessy committed:
[CL-3184] Add template serializer for initiatives-topics

adessy committed:
[CL-3184] Add template serializer for official feedbacks

adessy committed:
[CL-3184] Add template serializer for comments

adessy committed:
[CL-3184] Add template serializer for votes

adessy committed:
[CL-3184] Add template serializer for poll questions

adessy committed:
[CL-3184] Add template serializer for map config

adessy committed:
[CL-3184] Add template serializer for map layers

adessy committed:
[CL-3184] Add template serializer for legend items

adessy committed:
[CL-3184] Add template serializer for poll options

adessy committed:
[CL-3184] Add template serializer for poll responses

adessy committed:
[CL-3184] Add template serializer for poll-response options

adessy committed:
[CL-3184] Add template serializer for volunteering causes

adessy committed:
[CL-3184] Add template serializer for volunteers

adessy committed:
[CL-3184] Add template serializer for users

adessy committed:
[CL-3184] Add template serializer for custom fields

adessy committed:
[CL-3184] Add template serializer for admin publications

adessy committed:
[CL-3184] Add template serializer for projects

adessy committed:
[CL-3184] Add template serializer for phases

adessy committed:
[CL-3184] Add template serializer for ideas

adessy committed:
[CL-3184] Add template serializer for areas-projects

adessy committed:
[CL-3184] Add template serializer for unsubscription tokens

adessy committed:
[CL-3184] Add template serializer for project folders

adessy committed:
[CL-3200] Remove nested attributes

adessy committed:
[CL-3184] Fix serializer name

adessy committed:
areas-projects

adessy committed:
[CL-3184] Omit nil references from serialized records (template serializers)

adessy committed:
custom-field

adessy committed:
[CL-3184] Implement new tenant template serializer

... and adapt TenantTemplateService for compatibility

adessy committed:
[CL-3184] Refactor tenant_template_service.rb

- Rename variables for clarity
- Group instructions for better readability
- Remove unnecessary debugging logic

adessy committed:
[CL-3184] Simplify MultiTenancy::TenantTemplateService#required_locales

adessy committed:
[CL-3184] Remove MultiTenancy::TenantTemplateService#update_created_objects_ids

We can achieve the same with a hash with [] as default values.

adessy committed:
[CL-3184] Make the reference lookup fail if the object is missing

adessy committed:
[CL-3184] Use MultiTenancy::TenantTemplateService#multiloc where possible

adessy committed:
[CL-3184] Simplify MultiTenancy::TenantTemplateService#template_locales

adessy committed:
[CL-3184] Remove unnecessary mapping for legacy classes

Those classes are long gone and should not be found again in current templates.

adessy committed:
[CL-3184] Catch StandardError instead of Exception

adessy committed:
Merge branch 'master' into CL-3184/tenant-template-creation-prototype

adessy committed:
[CL-3184] Fix constant in tenant_serializer.rb

adessy committed:
[CL-3184] Refactoring MultiTenancy::TenantTemplateService#restore_multiloc_attribute

adessy committed:
[CL-3184] Add service to create s3-optimized templates

adessy committed:
[CL-3184] Document create_service.rb

adessy committed:
Merge branch 'master' into CL-3184/tenant-template-creation-prototype

adessy committed:
[CL-3184] Make copy_objects more flexible

adessy committed:
[CL-3184] Implement service to apply S3-based templates

adessy committed:
[CL-3184] Refactor TenantSerializer output to support new interfaces

In this commit, the output format of the TenantSerializer was updated to be compatible with the newly introduced MultiTenancy::Templates::CreateService and MultiTenancy::Templates::ApplyService. Previously, the serializer returned a serialized tenant that could be directly used with MultiTenancy::TenantTemplateService. However, now the output needs to be converted using MultiTenancy::Templates::TenantSerializer.format_for_tenant_template_service! before it can be used with the TenantTemplateService.

The updated format includes additional information about the original model identifiers, which is necessary for importing and copying template uploads stored in S3.

adessy committed:
[CL-3184] Disable some validations/callbacks when loading tenant templates

adessy committed:
[CL-3184] Replace the old tenant serializer with the new implementation

adessy committed:
[CL-3184] Adapt/refactor the logic to get the list of templates

We had to adapt this code because "external" templates are not stored as a single yaml file anymore since they also include all the uploads (and possibly a metadata.yml file in the future).

adessy committed:
[CL-3184] Extract template utils into a class

adessy committed:
[CL-3184] Rename TenantTemplateService to TenantDeserializer

The name is being changed because many responsibilities have been removed from this service. Its new focus will be on creating database records from serialized models.

adessy committed:
[CL-3184] Micro-refactoring

adessy committed:
[CL-3184] Extract template utils into a class

adessy committed:
[CL-3184] Move TenantDeserializer to the Templates namespace

adessy committed:
[CL-3184] Fix rubocop offenses

adessy committed:
[CL-3184] Adapt the ApplyService to support both internal and external templates

adessy committed:
[CL-3184] Add a prefix to the templates

The prefixes will enable us to categorize templates into different groups (as was already the case with the previous implementation of tenant templates):
- release
- test
- ...

adessy committed:
[CL-3184] Fix method name

adessy committed:
[CL-3184] Adapt ApplyTenantTemplateJob to use the tenant template system

adessy committed:
[CL-3184] Refactor tenant_serializer.rb

adessy committed:
[CL-3184] Replace YAML.load_file with YAML.load

The current version of YAML.load_file does not seem to support aliases, which are necessary for tenant templates.

adessy committed:
[CL-3184] Remove debug comment

adessy committed:
[CL-3184] Disable validations for file presence when loading tenant templates.

adessy committed:
[CL-3184] Fix InitiativeTopic serializer

InitiativesTopic's don't have created_at and updated_at timestamps.

adessy committed:
[CL-3184] Fix tenant serializer spec

Use the right template format when calling the deserializer.

adessy committed:
[CL-3184] Ignore rubocop guard clause offense

adessy committed:
[CL-3184] Fix and refactor template utils

It was referring to resolve_template which was out of scope.

adessy committed:
[CL-3184] Fix utils_spec.rb

adessy committed:
[CL-3184] Adapt tenant_deserializer.rb to make it work properly when running tests

When running tests, upload attributes can be File objects instead of strings.

adessy committed:
[CL-3184] Remove deprecated TenantDeserializer#resolve_and_apply_template method

This one should be removed as it is the responsibility of MultiTenancy::Templates::ApplyService to apply the templates.

adessy committed:
[CL-3184] Document TenantDeserializer

adessy committed:
[CL-3184] Remove unused parameter

adessy committed:
[CL-3184] Make template_bucket parameter optional

But if it is not specified some utils method will fail.

adessy committed:
[CL-3184] Remove usages of legacy classes

adessy committed:
[CL-3184] Make the "tenant_bucket" and "template_bucket" parameters optional in ApplyService

These parameters are not necessary when applying an internal template. However, the application of  external templates will fails if they are missing.

adessy committed:
[CL-3184] Fix tenant hostnames

Underscores are not allowed in host names.

adessy committed:
[CL-3184] Specify admin_publication_attributes explicitly in internal templates

Templates containing publications (folders or projects) must explicitly specify the admin publication attributes for each publication. This can be done either with nested attributes or by including the AdminPublication in the template. This is necessary because the :set_admin_publication hook in Folder and Project is disabled when loading a template, which means that the admin publication will not be created automatically.

adessy committed:
[CL-3184]  Raise an error when trying to fetch models of a non-existing template

adessy committed:
[NO-TICKET] Remove call to deprecated _pry_

> warning: _pry_ is deprecated, use pry_instance instead

adessy committed:
[NO-TICKET] Fix "repeat last command" with pry in Rails console

`pry_instance.run_command` didn't seem to work properly with the rails console.

adessy committed:
[CL-3184] Fix ApplyService specs

adessy committed:
[CL-3184] Fix rubocop offenses

adessy committed:
[CL-3184] Fix rubocop offenses

adessy committed:
[CL-3184] Fix key transform in Templates::ApplyService

adessy committed:
[CL-3184] Refactor MultiTenancy::Templates::CreateService#initialize signature

adessy committed:
[CL-3184] Rework template rake tasks

Adapt template rake tasks to the interface(s) of the new template system.

adessy committed:
[CL-3184] Use the computed release prefix as default instead of the hardcoded prefix

adessy committed:
[CL-3184] Fix rubocop offenses

adessy committed:
[CL-3184] Make MultiTenancy::Templates::ApplyService#copy_s3_files private

kogre reviewed (approved):
I scrolled through the code, trying to understand the high-level structure, and skimming certain sections that looked more complicated.

Afraid I can't really add anything useful without spending a lot of time, playing interpreter, not sure that's worth it. Don't have anything to remark on the code I did read more in depth.

In general, the code looks tidy, like the new serializer format, like the way you added comments. Main concern is bugs (as it's a lot of changes), so hope you can test enough.

adessy committed:
Merge branch 'master' into CL-3184/tenant-template-creation-prototype

# Conflicts:
#	back/engines/commercial/multi_tenancy/app/services/multi_tenancy/tenant_template_service.rb
#	back/engines/commercial/multi_tenancy/lib/tasks/core/add_builtin_custom_fields.rake
#	back/engines/commercial/multi_tenancy/lib/tasks/core/create_initiative_statuses.rake
#	back/engines/commercial/multi_tenancy/lib/tasks/core/setup_and_support.rake
#	back/engines/commercial/multi_tenancy/lib/tasks/core/templates.rake
#	back/engines/commercial/multi_tenancy/spec/services/multi_tenancy/templates/tenant_serializer_spec.rb

adessy committed:
[CL-3184] Fix tenant_serializer.rb

Some CustomFieldOption's and TextImage's were not serialized.

adessy committed:
[CL-3184] Fix apply_service_spec.rb

You should not have to specify the template bucket if you are just using internal templates.

adessy committed:
[CL-3184] Fix s3 prefix for templates in various places

adessy committed:
[CL-3184] Make MultiTenancy::Templates::Utils.user_locales compatible with both formats of serialized models

This commit makes modifications to MultiTenancy::Templates::Utils.user_locales to ensure that it works seamlessly with both formats of serialized models. The two formats supported are:

Format 1:
```
{
  'models': {
    class_name (in snakecase): [
      {attrs},
      {attrs},
      ...
    ]
  }
}
```

Format 2 (new format):
{
  'models': {
    class: {
      id: {attrs}
    }
  }
}

adessy committed:
[CL-3411] Fix rubocop offenses

adessy committed:
[CL-3411] Fix apply_service_spec.rb

You should not have to specify the template bucket if you are just using internal templates.
